File: taurus-protect-sdk-typescript/src/mappers/business-rule.ts
/**
 * Business rule mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { BusinessRule, BusinessRuleCurrency } from '../models/business-rule';
import { safeInt, safeString } from './base';

/**
 * Maps a currency DTO to a BusinessRuleCurrency model.
 */
export function businessRuleCurrencyFromDto(dto: unknown): BusinessRuleCurrency | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }
  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    symbol: safeString(d.symbol),
    name: safeString(d.name),
  };
}

/**
 * Maps a business rule DTO to a BusinessRule model.
 *
 * Handles the tenantId conversion from string to number to match the Java SDK.
 * Also normalizes entityID/entityId field name differences.
 */
export function businessRuleFromDto(dto: unknown): BusinessRule | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;

  return {
    id: safeString(d.id),
    tenantId: safeInt(d.tenantId),
    currency: safeString(d.currency),
    walletId: safeString(d.walletId),
    addressId: safeString(d.addressId),
    ruleKey: safeString(d.ruleKey),
    ruleValue: safeString(d.ruleValue),
    ruleGroup: safeString(d.ruleGroup),
    ruleDescription: safeString(d.ruleDescription),
    ruleValidation: safeString(d.ruleValidation),
    currencyInfo: businessRuleCurrencyFromDto(d.currencyInfo),
    entityType: safeString(d.entityType),
    entityId: safeString(d.entityID ?? d.entityId),
  };
}

/**
 * Maps an array of business rule DTOs to BusinessRule models.
 */
export function businessRulesFromDto(dtos: unknown[] | null | undefined): BusinessRule[] {
  if (!dtos) {
    return [];
  }
  const result: BusinessRule[] = [];
  for (const dto of dtos) {
    const mapped = businessRuleFromDto(dto);
    if (mapped !== undefined) {
      result.push(mapped);
    }
  }
  return result;
}

File: taurus-protect-sdk-python/taurus_protect/models/business_rule.py
"""Business rule models for Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field

from taurus_protect.models.currency import Currency
from taurus_protect.models.pagination import CursorPage


class BusinessRule(BaseModel):
    """Business rule for transaction validation and operational constraints."""

    id: str = Field(default="", description="Unique rule identifier")
    tenant_id: int = Field(default=0, description="Tenant ID")
    currency: Optional[str] = Field(default=None, description="Currency code")
    wallet_id: Optional[str] = Field(default=None, description="Wallet ID")
    address_id: Optional[str] = Field(default=None, description="Address ID")
    rule_key: Optional[str] = Field(default=None, description="Rule key/name")
    rule_value: Optional[str] = Field(default=None, description="Rule value/setting")
    rule_group: Optional[str] = Field(default=None, description="Rule group/category")
    rule_description: Optional[str] = Field(default=None, description="Human-readable description")
    rule_validation: Optional[str] = Field(default=None, description="Validation pattern")
    entity_type: Optional[str] = Field(
        default=None, description="Entity type (global, currency, wallet, address)"
    )
    entity_id: Optional[str] = Field(default=None, description="Entity ID")
    currency_info: Optional[Currency] = Field(default=None, description="Currency metadata")

    model_config = {"frozen": True}


class BusinessRuleResult(BaseModel):
    """
    One page of business rules.

    Continue with ``cursor=result.page.next_cursor`` while ``result.page.has_more``.
    """

    rules: List[BusinessRule] = Field(default_factory=list)
    page: CursorPage = Field(default_factory=CursorPage)

    model_config = {"frozen": True}
